TulparAudio/internal/source/SourceCollection.cpp

```diff
@@ -0,0 +1,1097 @@
+                break;
+            }
+            case AL_STATIC:
+            {
+                type = Source::Type::Static;
+                break;
+            }
+            case AL_STREAMING:
+            {
+                type = Source::Type::Streaming;
+                break;
+            }
+            default:
+            {
+                type = Source::Type::Unknown;
+                break;
+            }
+        }
+    }
+    else
+    {
+        LOG_AUDIO->Warning("Source #{}: get type: {:#x}", source, alErr);
+    }
+
+    return type;
+}
+
+bool SourceCollection::IsSourceRelative(SourceHandle source) const
+{
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALint result = AL_FALSE;
+
+    alGetSourcei(static_cast<ALuint>(source), AL_SOURCE_RELATIVE, &result);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        result = AL_FALSE;
+        LOG_AUDIO->Warning("Source #{}: get relative: {:#x}", source, alErr);
+    }
+
+    return result == AL_TRUE;
+}
+
+bool SourceCollection::SetSourceRelative(SourceHandle source, bool flag)
+{
+    LOG_AUDIO->Debug("Source #{}: set relative {}", source, flag);
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcei(static_cast<ALuint>(source), AL_SOURCE_RELATIVE, (flag ? AL_TRUE : AL_FALSE));
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set relative: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+bool SourceCollection::IsSourceLooping(SourceHandle source) const
+{
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALint result = AL_FALSE;
+
+    alGetSourcei(static_cast<ALuint>(source), AL_LOOPING, &result);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        result = AL_FALSE;
+        LOG_AUDIO->Warning("Source #{}: get looping: {:#x}", source, alErr);
+    }
+
+    return result == AL_TRUE;
+}
+
+bool SourceCollection::SetSourceLooping(SourceHandle source, bool flag)
+{
+    LOG_AUDIO->Debug("Source #{}: set looping {}", source, flag);
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcei(static_cast<ALuint>(source), AL_LOOPING, (flag ? AL_TRUE : AL_FALSE));
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set looping: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+float SourceCollection::GetSourcePitch(SourceHandle source) const
+{
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALfloat result = 0.0f;
+
+    alGetSourcef(static_cast<ALuint>(source), AL_PITCH, &result);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        result = -1.0f;
+        LOG_AUDIO->Warning("Source #{}: get pitch: {:#x}", source, alErr);
+    }
+
+    return result;
+}
+
+bool SourceCollection::SetSourcePitch(SourceHandle source, float value)
```

Comment:
Shouldn't it be const and static?

Reply:
Answered in https://github.com/Godlike/Tulpar/pull/6#discussion_r244268061